src/libcmis/oauth2-providers.cxx

```diff
@@ -292,18 +299,22 @@ int OAuth2Providers::parseResponse ( const char* response, string& post, string&
             xmlChar* action = xmlTextReaderGetAttribute( reader, 
                                                          BAD_CAST( "action" ));
 
-            // GDrive pin code page contains 3 forms.
-            // First form resends pin code, we need to omit its input fields.
-            // Also we mustn't parse action field from the 'select challenge' form.
+            // GDrive pin code page contains many forms.
+            // We have to parse only the form with pin field.
             if ( action != NULL )
             {
-                if ( strcmp((char*)action, "/signin/challenge") != 0
-                   && strcmp((char*)action, "/signin/selectchallenge") != 0
-                   && xmlStrlen(action) > 0 )
+                bool bChallengePage = ( xmlStrlen( action ) > 0 )
```

Comment:
I don't see any point in the xmlStrlen check. If action is "", the comparison will fail anyway.